### ISSUE TITLE
Allow excluding transitive dependencies

### ISSUE THREAD

**Issue by kamalmarhubi** [pubref/rules_maven#2]
This would be on a per-dep basis, achieving something like gradle's

```gradle
// From https://gist.github.com/xian/05c4f27da6d4156b9827842217c2cd5c
dependencies {
    compile("org.robolectric:robolectric:3.3") {
        // we don't need these MavenDependencyResolver in a hermetic build
        exclude group: 'org.apache.maven', module: 'maven-ant-tasks'
        exclude group: 'org.apache.ant', module: 'ant'
    }
```

This might require changing the `deps` attribute to a dict, or allowing some dict elements in the list, eg

```bzl
load("@org_pubref_rules_maven//maven:rules.bzl", "maven_repository")                                                                                                                                               

maven_repository(
    name = "robolectric",
    deps = [
        {
            "artifact": "org.robolectric:robolectric:3.3",
            "exclude": [
                "org.apache.maven:maven-ant-tasks",
                "org.apache.ant:ant",
            ],
        },
        "org.robolectric:shadows-core:3.3",
        "org.robolectric:shadows-httpclient:3.3",
        "org.robolectric:shadows-maps:3.3",
        "org.robolectric:shadows-multidex:3.3",
        "org.robolectric:shadows-play-services:3.3",
        "org.robolectric:shadows-support-v4:3.3",                                                                                                                                                                  
        "org.robolectric:android-all:7.1.0_r7-robolectric-0",
    ],
)
```

**Comment by kamalmarhubi** [comment]
Looks like the `attr` module won't allow mixing strings and dicts. Will consider alternatives.

**Comment by pcj** [comment]
Could add another attr to support these more complex dependency declarations.

**Comment by pcj** [comment]
AFAIK bazel does not support creating attributes that have internal structure like the form you are proposing:  

What about an `exclude` [string_list_dict](https://bazel.build/versions/master/docs/skylark/lib/attr.html#string_list_dict) attr that looks like:

```python
maven_repository(
    name = 'robolectric',
    deps = [
        'org.robolectric:robolectric:3.3',
        'org.robolectric:shadows-core:3.3',
        'org.robolectric:shadows-httpclient:3.3',
        'org.robolectric:shadows-maps:3.3',
        'org.robolectric:shadows-multidex:3.3',
        'org.robolectric:shadows-play-services:3.3',
        'org.robolectric:shadows-support-v4:3.3',
        'org.robolectric:android-all:7.1.0_r7-robolectric-0',
    ],
    exclude = {
        'org.robolectric:robolectric': [
            'org.apache.maven:maven-ant-tasks',
            'org.apache.ant:ant',
        ],
    },
    ...
)
```

In this form you have to provide the artifact name as the key to associate the excluded dependencies.

**Comment by kamalmarhubi** [comment]
I implemented this in a less nice way, where I use `string_list_dict` for `deps` and it looked like this:

```bazel
maven_repository(                                                                                                                                                                                                  
    name = 'robolectric',
    deps = {
        'org.robolectric:android-all:7.1.0_r7-robolectric-0': [],
        'org.robolectric:robolectric:3.3': [
            "org.apache.maven:maven-ant-tasks",
            "org.apache.ant:ant",
        ]
    },
)
```

I'll take your suggested layout and make a PR soon.

**Comment by pcj** [comment]
Hi Kamal,

I worked on this a bit this morning and pushing it up now. Link to the PR #3 is provided although I wouldn't merge it as-is.  

It simply uses a `java_test` to trigger the upstream `maven_repository` rule to test the `exclude` attribute, recognizing that's pretty contrived for an android based project.

If you want to expand on this to tune it up and submit a PR that addresses your use case, great.  Moreover, if you can create a functioning robolectric test, it would make an awesome example for this repo.  I understand there is talk of an `android_robolectric_test` rule from https://github.com/bazelbuild/bazel/issues/2080 and https://github.com/bazelbuild/bazel/issues/903.  Sounds somewhat daunting.  

Interested in your thoughts and where you're going with bazel + robolectric + rules_maven.

**Comment by kamalmarhubi** [comment]
@pcj sorry for the delay. What would be necessary to merge this?

> It simply uses a `java_test` to trigger the upstream `maven_repository` rule to test the `exclude` attribute, recognizing that's pretty contrived for an android based project.

Not sure I follow you!

As for my use case right now, I just want to run tests that use Android libraries. I'm not building an application at this point.

**Comment by pcj** [comment]
No worries.  I can probably tidy and up and merge it over the weekend.

**Comment by pcj** [comment]
Closed by #5
